numpy/core/src/common/npy_cpu_features.c.src

```diff
@@ -94,6 +112,135 @@ npy_cpu_dispatch_list(void)
+{
+#if !defined(NPY_DISABLE_OPTIMIZATION) && NPY_WITH_CPU_BASELINE_N > 0
+    char baseline_failure[sizeof(NPY_WITH_CPU_BASELINE) + 1];
+    char *fptr = &baseline_failure[0];
+
+    #define NPY__CPU_VALIDATE_CB(FEATURE, DUMMY)                  \
+        if (!npy__cpu_have[NPY_CAT(NPY_CPU_FEATURE_, FEATURE)]) { \
+            const int size = sizeof(NPY_TOSTRING(FEATURE));       \
+            memcpy(fptr, NPY_TOSTRING(FEATURE), size);            \
+            fptr[size] = ' '; fptr += size + 1;                   \
+        }
+    NPY_WITH_CPU_BASELINE_CALL(NPY__CPU_VALIDATE_CB, DUMMY) // extra arg for msvc
+    *fptr = '\0';
+
+    if (baseline_failure[0] != '\0') {
+        *(fptr-1) = '\0'; // trim the last space
+        PyErr_Format(PyExc_RuntimeError,
+            "The NumPy library was compiled with a minimal set of required optimizations: \n"
+            "(" NPY_WITH_CPU_BASELINE ") but your machine doesn't support:\n(%s).\n",
+            baseline_failure
+        );
+    }
+#endif
+}
+
+static void
+npy__cpu_try_disable_env(void)
+{
+    char *disenv = getenv("NPY_DISABLE_CPU_FEATURES");
+    if (disenv == NULL || disenv[0] == 0) {
+        return;
+    }
+#if !defined(NPY_DISABLE_OPTIMIZATION) && NPY_WITH_CPU_DISPATCH_N > 0
+    #define NPY__MAX_VAR_LEN 1024 // More than enough for this era
+    size_t var_len = strlen(disenv) + 1;
+    if (var_len > NPY__MAX_VAR_LEN) {
+        PyErr_Format(PyExc_RuntimeError,
+            "Maximum acceptable length of NumPy's environment variable "
+            "'NPY_DISABLE_CPU_FEATURES' is %d",
+            NPY__MAX_VAR_LEN - 1
+        );
+        return;
+    }
+    char disable_features[NPY__MAX_VAR_LEN];
+    memcpy(disable_features, disenv, var_len);
+
+    //comma and space including (htab, vtab, CR, LF, FF)
+    const char *delim = ", \t\v\r\n\f";
+    char *feature = strtok(disable_features, delim);
+    while (feature) {
+        #define NPY__CPU_ENV_ERROR_HEAD \
+            "Invalid use of environment variable 'NPY_DISABLE_CPU_FEATURES',\nsince "
+
+        if (npy__cpu_baseline_fid(feature) > 0) {
+            PyErr_Format(PyExc_RuntimeError,
+                NPY__CPU_ENV_ERROR_HEAD
+                "CPU feature '%s' is part of the minimal set of required optimizations:\n"
+                "(" NPY_WITH_CPU_BASELINE ").\n",
+                feature
+            );
+            break;
+        }
+        int feature_id = npy__cpu_dispatch_fid(feature);
+        if (feature_id == 0) {
+            PyErr_Format(PyExc_RuntimeError,
+                NPY__CPU_ENV_ERROR_HEAD
+                "CPU feature '%s' is not part of dispatched set of additional optimizations:\n"
+                "(" NPY_WITH_CPU_DISPATCH ").\n",
+                feature
+            );
+            break;
+        }
+        if (!npy__cpu_have[feature_id]) {
+            PyErr_Format(PyExc_RuntimeError,
+                NPY__CPU_ENV_ERROR_HEAD
+                "CPU feature '%s' is not supported by your machine.\n",
```

Comment:
if a non CPU supported feature is disabled, would it be better to warn than error in this case ?

Reply:
agreeed

Reply:
agreed and done